src/main.rs

```diff
@@ -223,7 +298,14 @@ fn watch(exercises: &[Exercise], verbose: bool) -> notify::Result<()> {
                         let filepath = b.as_path().canonicalize().unwrap();
                         let pending_exercises = exercises
                             .iter()
-                            .skip_while(|e| !filepath.ends_with(&e.path));
+                            .filter(|e| filepath.ends_with(&e.path))
```

Comment:
I'm inclined to keep `skip_while` here, so in the case of a jump, say, from exercise7 to exercise14, the user can still follow the logical thread of exercises from exercise14. In this way the `chain`ed iterator would work as a second pass just in case you skipped something.

I like that, for a "normal" flow, this does solve #465. But note that if the user skips some exercises, creating an extraneous file will again show some weird behaviour (we will take them back to the first skipped exercise). This is tricky to solve in a stateless fashion. If we do want to solve this case, maybe we could refactor and repurpose `failed_exercise_hint`.

Reply:
Hi @jrvidal thanks for the very thorough review! I struggled with this update, but I see your point about somehow keeping `skip_while` for the maintenance of a logical thread. However, it should be mentioned that keeping `skip_while` and `chaining` may lead to the `pending_exercises` iterator containing duplicates. In my naive thinking, going back to the skipped exercise sounded preferable to doing duplicate compilations. But like you said, it's hard to fix this issue with maintaining some actual state (not just whether the "I AM NOT DONE" string is removed from the files.)
Do you know of a cleaner way of keeping `skip_while` without introducing duplicates?

Reply:
Do those duplicates have an observable effect for the end user? I suspect not (though I haven't given it much thought).

Reply:
Yes. The updated file does not get compiled again when the following is used:
```
let pending_exercises = exercises
    .iter()
    .skip_while(|e| !e.path.ends_with(&filepath))
    // .filter(|e| filepath.ends_with(&e.path))
    .chain(
        exercises
            .iter()
            .filter(|e| !e.looks_done() && !filepath.ends_with(&e.path))
);
```
Perhaps, I missing something here?

Reply:
But this iterator is passed to `verify()` which does not necessarily compile and run every single exercise within it. It stops as soon as it finds one that is not actually done. So I'm having a hard time imagining a case where we end up actually compiling twice the same file.

Reply:
The iterator itself will contain the duplicate, but like you said `verify` does stop upon encountering a failed exercise. So, it may not matter at all if the iterator contains duplicates.
However, the chaining using `skip_while` is not working at all, since the updated file does not get compiled. Please feel free to toggle between `filter` and `skip_while` and let me know if you can observe when I am referring to.

Reply:
I do see weird behaviour toggling between the commented out version and the current one. However,  when I suggested keeping `skip_while`, I was referring to the old version:
```rust
.skip_while(|e| !filepath.ends_with(&e.path))
```
I think it works fine in this way.

Reply:
Keeping that line was the cause of the bug reported in the target issues. I have explained why in issue #585.

When you have the original line, it makes it so that if you skip the order given by the `watch` command and solve the last problem (which is currently `exercises/conversions/from_str.rs`), the program exits and claims that you completed all the exercises, even when you've still got pending exercises.

Please feel free to try and reproduce that bug by invoking `watch` and then finishing `exercises/conversions/from_str.rs`. Regardless of whether or not you've got other pending exercises, the program will exit and print out an erroneous message you've finished all exercises.

Reply:
Sorry, I think I might have been too terse, or maybe it's just my obtuseness :sweat_smile:

What I'm suggesting is something along the lines of:
```rust
// Original line, unchanged
.skip_while(|e| !filepath.ends_with(&e.path))
// NEW
.chain(
  exercises
    .iter()
    .filter(|e| /* your thing */)
)
```

AFAICT this:

* Avoids #585 by doing a second pass and looking for unfinished exercises.
* Avoids #465 by using this second pass as fallback when we get an event about an extraneous file.
* Keeps the current behavior where we progress through the exercises list in order (at least, by default).
* Never verifies an exercise twice.

Reply:
I think the weird behavior I was observing with `skip_while` is due to how I set up the condition: `.skip_while(|e| !e.path.ends_with(&filepath))`. It should have been `.skip_while(|e| !filepath.ends_with(&e.path))`. `filepath` is the canonical path. That's why the least recently modified file was not being checked. I think that's what I was missing. Thank you!

While I am more than fine with this current behavior, I think this behavior _does_ verify files twice. You can see this by updating an already solved exercises that precedes the one `rustlings watch` is suggesting you solve. For example, if you have already solved the first 5 exercises and you're being prompted to solve the 6th. However, you decide, for whatever reason, to go back to exercise 1 and update it. Upon updating exercise 1, `watch` will verify exercises 1 through 6 again. I am happy to demo this over slack or Zoom, whenever time allows. My description of the behavior may not be clear enough.

In any case, like I said, I am fine with this behavior but I do want to make sure we're on the same page and are seeing the same behavior reflected.

Reply:
Ah, I see, yes, that is a non ideal UX. <sub>Though note that my concern with "double verification" was more about not observing rustlings verify the same exercise twice on the same verification batch.</sub> If a user wanted to re-evaluate their solution, they'd need to re-add `I AM NOT DONE`, which might not be entirely obvious :confused:

I think that we can live with that, it is after all the _current_ behavior. I think the `chain`ed fallback is an improvement anyway. @fmoko Thoughts?

Reply:
@fmoko, the current code is in the state @jrvidal suggested. It works fine, to me. It can be merged. If this comes back again in the form of an issue, we could toggle back to the `filter` solution as an alternative.

Thanks,
Abdou